store/postgres/migrations/2019-03-07-171355_update_history_storage/up.sql

```diff
@@ -0,0 +1,178 @@
+        operation_type := 1;
+        event_source := NEW.event_source;
+        subgraph := OLD.subgraph;
+        entity := OLD.entity;
+        entity_id := OLD.id;
+        data_before := OLD.data;
+        data_after := NEW.data;
+    ELSIF (TG_OP = 'DELETE') THEN
+        operation_type := 2;
+        event_source := current_setting('vars.current_event_source', TRUE);
+        subgraph := OLD.subgraph;
+        entity := OLD.entity;
+        entity_id := OLD.id;
+        data_before := OLD.data;
+        data_after := NULL;
+    ELSE
+        RAISE EXCEPTION 'unexpected entity row operation type, %', TG_OP;
+    END IF;
+
+    -- We do not need any history for the 'subgraphs' subgraph nor
+    -- do we need to record changes that do not change anything
+    IF subgraph = 'subgraphs' OR data_before = data_after THEN
+        RETURN NULL;
+    END IF;
+
+    IF event_source = 'REVERSION' THEN
+        is_reversion := TRUE;
+    END IF;
+
+    SELECT id INTO event_id
+    FROM event_meta_data
+    WHERE db_transaction_id = txid_current();
+
+    new_event_id := null;
+
+    IF event_id IS NULL THEN
+        -- Log information on the postgres transaction for later use in
+        -- revert operations
+        INSERT INTO event_meta_data
+            (db_transaction_id, db_transaction_time, source)
+        VALUES
+            (txid_current(), statement_timestamp(), event_source)
+        RETURNING event_meta_data.id INTO new_event_id;
+    END IF;
+
+    -- Log row metadata and changes, specify whether event was an original
+    -- ethereum event or a reversion
+    INSERT INTO entity_history
+        (event_id, entity_id, subgraph, entity,
+         data_before, data_after, reversion, op_id)
+    VALUES
+        (COALESCE(new_event_id, event_id), entity_id, subgraph, entity,
+         data_before, data_after, is_reversion, operation_type);
+    RETURN NULL;
+END;
+$$ LANGUAGE plpgsql;
+
+-- Create a view to get entity_history by source ordered by event_id desc
+-- We use this both in revert_block and in the Rust code in
+-- store_events.get_revert_event
+create or replace view entity_history_with_source as
+select
+  h.subgraph,
+  h.entity,
+  h.entity_id,
+  h.data_before,
+  h.op_id,
+  m.source
+from entity_history h, event_meta_data m
+where h.event_id = m.id
+order by h.event_id desc;
+
+-- Rewrite revert_block to be selfcontained and to get all the information
+-- needed for reversion in one query
+CREATE OR REPLACE FUNCTION
+  revert_block(block_to_revert_hash VARCHAR,
+               subgraph_id VARCHAR)
+    RETURNS VOID AS
+$$
+DECLARE
+    history RECORD;
+BEGIN
+    -- Revert all relevant events
+    FOR history IN
+        -- Get all history entries associated with the given block
+    SELECT
+        h.entity,
+        h.entity_id,
+        h.data_before,
+        h.op_id
+    FROM entity_history_with_source h
+    WHERE h.source = block_to_revert_hash
+      AND h.subgraph = subgraph_id
+    LOOP
```

Comment:
I was going to say "these were ordered before" and then I realized `entity_history_with_source` _is_ pre-ordered. It would be nice to have a comment here that expresses that _or_ to rename `entity_history_with_source to `ordered_entity_history_with_source` to make this obvious.

Reply:
Ok. I'll add a comment, both here and where the view is defined.